### PR TITLE
Adds hasJoinedBefore and PlayerFirstJoinEvent. SPOUT-220

### PR DESCRIPTION
**Pull request by Amperial**

### CODE REVIEW

**Comment by lukespragg** [comment]
In addition to the formatting comments I made, you'd need to sign-off on your commit.

Sign-off, verifying agreement with the license terms.

A commit with no detailed description can be created using:
git commit -s -m "A brief description of the change"

**Comment by Amperial** [comment]
Fixed.

**Comment by zml2008** [comment]
Having a separate event is stupid. If statements exist for a reason.

**Comment by Amperial** [comment]
It seems like @kitskub @thehutch and Ian think it would be better to have a separate one, I kinda agree.
I can make it one event again if you guys want though?

**Comment by zml2008** [comment]
The subclass adds no methods, only exists for `event instance of PlayerFirstJoin`, which can be done just as well with `!player.hasJoinedBefore()`

**Comment by Amperial** [comment]
What if I added a get/set welcome message to the first join event, that gets sent right before the join message?
That way it would actually serve a purpose, and servers wouldn't need a plugin to take care of that.

**Comment by Grinch** [comment]
Vanilla isn't set out to replace plugins, just to provide an extensible API. This event (in my opinion) isn't really needed as they can simply do a `if (!hasJoinedBefore)` check during the PlayerJoinEvent. :/

**Comment by kitskub** [comment]
As long hasJoinedBefore is false during the event, then no need for new event.

File: engine/src/main/java/org/spout/engine/entity/SpoutPlayer.java [hunk 1]
```diff
@@ -89,8 +90,12 @@ public SpoutPlayer(Engine engine, Class<? extends PlayerNetworkComponent> networ
 		copySnapshot();
 	}
 
+	/**
+	 * Only used when creating a new SpoutPlayer, hence firstLogin being set to true.
+	 */
 	public SpoutPlayer(Engine engine, Class<? extends PlayerNetworkComponent> network, String name, Transform transform) {
 		this(engine, network, name, transform, null, (byte[]) null, (Class<? extends Component>[]) null);
+	    firstLogin = true;
```
Comment (kitskub) [inline engine/src/main/java/org/spout/engine/entity/SpoutPlayer.java]: Formatting...

File: api/src/main/java/org/spout/api/entity/Player.java [hunk 2]
```diff
@@ -68,6 +68,13 @@
 	public boolean isOnline();
 
 	/**
+	 * Gets if the player has joined before
+	 *
```
Comment (kitskub) [inline api/src/main/java/org/spout/api/entity/Player.java]: Since you have to edit the commit for formatting below, can you remove this space? Thanks!

Comment (lukespragg) [inline api/src/main/java/org/spout/api/entity/Player.java]: Trailing whitespace, not the new line. ^

File: engine/src/main/java/org/spout/engine/entity/SpoutPlayer.java (hunk 1 above)
Comment (lukespragg) [inline engine/src/main/java/org/spout/engine/entity/SpoutPlayer.java]: By formatting, he means we use tabs, not spaces.

**Comment by Amperial** [comment]
Yeah i'll fix it tonight, sorry :P
Usually I just auto format and it removes the spaces, but that would change a bunch of other stuff so I had to do it manually.

File: engine/src/main/java/org/spout/engine/entity/SpoutPlayer.java (hunk 1 above)
Comment (kitskub) [inline engine/src/main/java/org/spout/engine/entity/SpoutPlayer.java]: Si

File: api/src/main/java/org/spout/api/entity/Player.java (hunk 2 above)
Comment (kitskub) [inline api/src/main/java/org/spout/api/entity/Player.java]: :dart:

**Comment by Amperial** [comment]
Fixed the two formatting errors.

**Comment by kitskub** [comment]
One last small thing (sorry!), can you edit the commit description to not include the " and PlayerFirstJoinEvent"?

Thanks! Really! That's the only thing I see!

**Comment by Amperial** [comment]
Alright, done :) @kitskub

**Comment by kitskub** [comment]
thanks!

**Pull request merged**